// farm-section/++/farm-section-bis.js
/*! farm_section_bis v0.00.0 | (c) 2015 SAM Staff, Inc. | 

           					   \|||/
                               (o o)
                      ------ooO-(_)-Ooo------

*/
// http://stackoverflow.com/questions/13673346/supporting-both-commonjs-and-amd

(function(root, factory) {
    if (typeof define === 'function' && define.amd) {
        define(['jquery'], factory);
    } else if (typeof exports === 'object') {
        module.exports = factory(jQuery);
    } else {
        root.farm_section_bis = factory(jQuery);
    }
}(this, function($) {
    'use strict';

    var commands = {
        unnamed             :unnamed,
        //
        //on                  :on,
        move                :move,
        re_move             :re_move,
        change              :change,
        show                :show,
        animation_end       :animation_end
    };

    var that;

    var farm_section_bis = function() {

        if (typeof arguments[0] === 'string') {
            var args = Array.prototype.slice.call(arguments, 1);
            commands[arguments[0]].apply(this, args);
        }
        else {
            that = this;
            create_plugin.apply(this, arguments);
        }
        return this;
    };

    var defaults = {
            mode            :'',
            current         :'default',
            pageDefault     :'', //glob.o.mode === 3 ? '/tm' : '/lb',
            animation       :true,

            open            :function (){ return true; },
            close           :function (){ return true; },

            routes          :function ($data){ 
                                //routes($data); 
                                return true; 
                            },
            show            :function ($data){ 
                                //routes.show($data, null, false);

                                //Exclude from history
                                //routes.replace(url, null, false);
                                //routes.show(url, null, false, false);
                                //history.replaceState({}, document.title, '#' + url);
                                
                                return true; 
                            },
            animations      :function ($current, $next){

                                var $ani;
                                
                                if( $current === 'default'){

                                    $ani = 21;
                                }
                                else if( $current === 'prefs' || $next === 'prefs'){

                                    if( $current === 'prefs'){
                                        $ani = 4;
                                    }
                                    else{
                                        $ani = 3;
                                    }
                                }
                                else{
                                    $ani = 5;
                                }

                                return $ani; 
                            },

            parentLocationHash :function ($data){
                                //para que el hash se actualide desde el m1 al m0 (SAM)

                                //parent.location.hash = $data;
                                return true; 
                            },
            changeSection   :function ($data){
                                //para parsear sections en el m0 (SAM)

                                //console.log('changeSection mode', glob.o.mode);

                                //if($data == 'tm'){$data = 'iframe'}
                                return $data; 
                            },
            next   :function ($data){
                                //para añadir parametros cuando se ejecuta siguiente

                                return $data; 
                            },
    };

    function create_plugin (options) {

        //console.log('this: ',this);
        //console.log('that: ',that);
        //console.log('options: ',options);
        
        
        that.options = extend( defaults, options);
        var opts = that.options;


        //pendiente de quitar el jquery
        on();
    }

    /////////////////////////////////////////////////////////////////////////////////

    function unnamed () {
        var opts = that.options;
    };

    /////////////////////////////////////////////////////////////////////////////////

    function on (){

        $("body").on("click", "[farm-navi]", function(){
            var hr  = this.getAttribute('farm-navi');
                /*
                bl  = this.getAttribute('sas-blind'),
                cl  = this.getAttribute('sas-clear'),
                to  = this.getAttribute('sas-timeout'),
                st  = Number( this.getAttribute('sas-transition') );
                */

            //change( hr.replace('\/','').replace('\#','') );
            //move( hr, cl, to);

            move( hr );
        });
    };
 

    //function move ( nextp, clear, timeout ){
    function move ( nextp ){
        var opts = that.options;

        //console.log('******* farm-router move opts: ',opts)
        /*
        if(clear!=undefined){
            $('#'+clear+'-page')
                .empty()
                .html( '<span class="spinner-overlay"><p class="spinner">Please wait while we do what we do best.</p></span>' );
        }
        */

        var $current, $next;

        //console.log( opts.current.replace("/", "") );

        opts.current  = opts.current.replace("/", "");

        //console.log( 'opts: ', opts );


        $current = document.getElementById(opts.current);

        var np = nextp.substring(0,1) ==='/' ? nextp : '/'+nextp;    //nextp.replace('/', '');

        var gnext = get_next( np );

        //console.log('nextp', nextp)
        //console.log('gnext', gnext)

        if(gnext){

            // para cambiar ids

            //if(gnext == 'tm'){gnext = 'iframe'}

            var ngnext = opts.changeSection(gnext);

            //console.log('ngnext', ngnext)

            $next = document.getElementById(ngnext);

            if($next){

                opts.current = ngnext;
            }
        }

        console.log( 'current: ',$current );
        console.log( 'next: ',$next );

        end( $current, $next );
        opts.parentLocationHash(nextp);
        opts.routes(nextp); 
    };

    function get_next( nextp ){
        //console.log('nextp', nextp)

        var next = nextp.match(/^\/([^\/]*).*$/);

        //console.log('next', next)

        if(next[1].indexOf('?') > 0){
            var a = next[1].split("?");
            next[1] = a[0];
        }
        return next[1];
    };

    function end ( $current, $next ){

        var opts    = that.options;
 
        //console.log('####@@@@@@@@@@@#######', opts);
    
        //console.log($current);
        //console.log($next);


        if(opts.animation){

            var acurr = $current.getAttribute('id');
            var anext = $next.getAttribute('id');

            var $ani = opts.animations(acurr, anext);

            go( $current, $next, $ani );
        }
        else{

            opts.next($next);

            //$($current).removeAttr('farm-select');
            //$($next).attr('farm-select', true);

            if($current && $next){

                $current.removeAttribute('farm-select');
                $next.setAttribute('farm-select', true);
            }
        }
    };

    function re_move (){
        var opts = that.options;

        //console.log( window.location.hash )

        var url = window.location.hash.replace('#','').match(/(^\/.*)/);

        //var url = window.location.hash.match(/(^\/.*)/);

        //var url = window.location.hash;

        //console.log('url --> ', url);

        if( url ){

            //var u = url[1].replace('/','')

            move( url[1] );
        }
        else{

            move( opts.pageDefault );
        }
    }

    function change ( $next ){
        var opts = that.options;

        var cr = document.getElementById(opts.current);
        var nt = document.getElementById($next);

        //console.log( cr );
        //console.log( nt );

        end( cr, nt );

        opts.current = $next;
    };

    function show(url){
        var opts = that.options;

        opts.parentLocationHash(url);

        opts.show(url); 
    };

    ///////////////////////////////////////////////////
    ///////////////////////////////////////////////////


    function go( $current, $next, $ani ){

        var animation = $ani || 56;
        
        //console.log(animation);

        var aniEnd  = animEndEventName();

        //console.log('aniEnd -> '+aniEnd+' current -> '+idcurrent+' next -> '+idnext+' ani -> ' +st)

        /*
        if(aniEnd){

                ani ( aniEnd, $current, $next, animation );
        }else{
                aniOff ( $current, $next );
        }
        */


        ////////////pendiente de quitar jquery
        //////////////////////////////////////
        /*
        if(!aniEnd || aniEnd === 'MSAnimationEnd'){

                aniOff ( $current, $next );
        }else{
                ani ( aniEnd, $current, $next, animation );  
        }
        */


        aniOff ( $current, $next );
    };

    function animEndEventName() {

        var  s = document.body.style
            ,animationProperty =
                'WebkitAnimation' in s ? 'webkitAnimationEnd' :
                'OAnimation' in s ? 'oAnimationEnd' :
                'msAnimation' in s ? 'MSAnimationEnd':
                'animation' in s ? 'animationend': 'no-supported'
            ;

        //console.log( s ) ;
        //console.log( 'animationProperty -->', animationProperty );

        return animationProperty === 'no-supported' ? false : animationProperty;
    };

    function aniOff( $current, $next ) {

            //$current.removeClass( active );
            //$next.addClass( active );

            //$current.removeAttr('farm-select');
            //$next.attr('farm-select', true);

            $current.removeAttribute('farm-select');
            $next.setAttribute('farm-select', true);
    };

    function ani( animEndEventName, $current, $next, $ani ) {
        /*
        console.log('animation --> ',$ani);
        console.log('animation --> ',$current);
        console.log('animation --> ',$next);
        */
        var outClass, inClass;
        /*
        var s = document.body.style;
        var animationProperty =
                'WebkitAnimation' in s ? 'webkitAnimationEnd' :
                'OAnimation' in s ? 'oAnimationEnd' :
                'msAnimation' in s ? 'MSAnimationEnd':
                'animation' in s ? 'animationend': 'no-supported';
          
        console.log( 'animationProperty -->', animationProperty )

        animEndEventName = animationProperty;
        */

        var isAnimating = false;

        //if (target !== previous && !isAnimating) {
            isAnimating = true;
            

            switch( $ani ) {

                case 1:
                    outClass = 'pt-page-moveToLeft';
                    inClass = 'pt-page-moveFromRight';
                    break;
                case 2:
                    outClass = 'pt-page-moveToRight';
                    inClass = 'pt-page-moveFromLeft';
                    break;
                case 3:
                    outClass = 'pt-page-moveToTop';
                    inClass = 'pt-page-moveFromBottom';
                    break;
                case 4:
                    outClass = 'pt-page-moveToBottom';
                    inClass = 'pt-page-moveFromTop';
                    break;
                case 5:
                    outClass = 'pt-page-fade';
                    inClass = 'pt-page-moveFromRight pt-page-ontop';
                    break;
                case 6:
                    outClass = 'pt-page-fade';
                    inClass = 'pt-page-moveFromLeft pt-page-ontop';
                    break;
                case 7:
                    outClass = 'pt-page-fade';
                    inClass = 'pt-page-moveFromBottom pt-page-ontop';
                    break;
                case 8:
                    outClass = 'pt-page-fade';
                    inClass = 'pt-page-moveFromTop pt-page-ontop';
                    break;
                case 9:
                    outClass = 'pt-page-moveToLeftFade';
                    inClass = 'pt-page-moveFromRightFade';
                    break;
                case 10:
                    outClass = 'pt-page-moveToRightFade';
                    inClass = 'pt-page-moveFromLeftFade';
                    break;
                case 11:
                    outClass = 'pt-page-moveToTopFade';
                    inClass = 'pt-page-moveFromBottomFade';
                    break;
                case 12:
                    outClass = 'pt-page-moveToBottomFade';
                    inClass = 'pt-page-moveFromTopFade';
                    break;
                case 13:
                    outClass = 'pt-page-moveToLeftEasing pt-page-ontop';
                    inClass = 'pt-page-moveFromRight';
                    break;
                case 14:
                    outClass = 'pt-page-moveToRightEasing pt-page-ontop';
                    inClass = 'pt-page-moveFromLeft';
                    break;
                case 15:
                    outClass = 'pt-page-moveToTopEasing pt-page-ontop';
                    inClass = 'pt-page-moveFromBottom';
                    break;
                case 16:
                    outClass = 'pt-page-moveToBottomEasing pt-page-ontop';
                    inClass = 'pt-page-moveFromTop';
                    break;
                case 17:
                    outClass = 'pt-page-scaleDown';
                    inClass = 'pt-page-moveFromRight pt-page-ontop';
                    break;
                case 18:
                    outClass = 'pt-page-scaleDown';
                    inClass = 'pt-page-moveFromLeft pt-page-ontop';
                    break;
                case 19:
                    outClass = 'pt-page-scaleDown';
                    inClass = 'pt-page-moveFromBottom pt-page-ontop';
                    break;
                case 20:
                    outClass = 'pt-page-scaleDown';
                    inClass = 'pt-page-moveFromTop pt-page-ontop';
                    break;
                case 21:
                    outClass = 'pt-page-scaleDown';
                    inClass = 'pt-page-scaleUpDown pt-page-delay300';
                    break;
                case 22:
                    outClass = 'pt-page-scaleDownUp';
                    inClass = 'pt-page-scaleUp pt-page-delay300';
                    break;
                case 23:
                    outClass = 'pt-page-moveToLeft pt-page-ontop';
                    inClass = 'pt-page-scaleUp';
                    break;
                case 24:
                    outClass = 'pt-page-moveToRight pt-page-ontop';
                    inClass = 'pt-page-scaleUp';
                    break;
                case 25:
                    outClass = 'pt-page-moveToTop pt-page-ontop';
                    inClass = 'pt-page-scaleUp';
                    break;
                case 26:
                    outClass = 'pt-page-moveToBottom pt-page-ontop';
                    inClass = 'pt-page-scaleUp';
                    break;
                case 27:
                    outClass = 'pt-page-scaleDownCenter';
                    inClass = 'pt-page-scaleUpCenter pt-page-delay400';
                    break;
                case 28:
                    outClass = 'pt-page-rotateRightSideFirst';
                    inClass = 'pt-page-moveFromRight pt-page-delay200 pt-page-ontop';
                    break;
                case 29:
                    outClass = 'pt-page-rotateLeftSideFirst';
                    inClass = 'pt-page-moveFromLeft pt-page-delay200 pt-page-ontop';
                    break;
                case 30:
                    outClass = 'pt-page-rotateTopSideFirst';
                    inClass = 'pt-page-moveFromTop pt-page-delay200 pt-page-ontop';
                    break;
                case 31:
                    outClass = 'pt-page-rotateBottomSideFirst';
                    inClass = 'pt-page-moveFromBottom pt-page-delay200 pt-page-ontop';
                    break;
                case 32:
                    outClass = 'pt-page-flipOutRight';
                    inClass = 'pt-page-flipInLeft pt-page-delay500';
                    break;
                case 33:
                    outClass = 'pt-page-flipOutLeft';
                    inClass = 'pt-page-flipInRight pt-page-delay500';
                    break;
                case 34:
                    outClass = 'pt-page-flipOutTop';
                    inClass = 'pt-page-flipInBottom pt-page-delay500';
                    break;
                case 35:
                    outClass = 'pt-page-flipOutBottom';
                    inClass = 'pt-page-flipInTop pt-page-delay500';
                    break;
                case 36:
                    outClass = 'pt-page-rotateFall pt-page-ontop';
                    inClass = 'pt-page-scaleUp';
                    break;
                case 37:
                    outClass = 'pt-page-rotateOutNewspaper';
                    inClass = 'pt-page-rotateInNewspaper pt-page-delay500';
                    break;
                case 38:
                    outClass = 'pt-page-rotatePushLeft';
                    inClass = 'pt-page-moveFromRight';
                    break;
                case 39:
                    outClass = 'pt-page-rotatePushRight';
                    inClass = 'pt-page-moveFromLeft';
                    break;
                case 40:
                    outClass = 'pt-page-rotatePushTop';
                    inClass = 'pt-page-moveFromBottom';
                    break;
                case 41:
                    outClass = 'pt-page-rotatePushBottom';
                    inClass = 'pt-page-moveFromTop';
                    break;
                case 42:
                    outClass = 'pt-page-rotatePushLeft';
                    inClass = 'pt-page-rotatePullRight pt-page-delay180';
                    break;
                case 43:
                    outClass = 'pt-page-rotatePushRight';
                    inClass = 'pt-page-rotatePullLeft pt-page-delay180';
                    break;
                case 44:
                    outClass = 'pt-page-rotatePushTop';
                    inClass = 'pt-page-rotatePullBottom pt-page-delay180';
                    break;
                case 45:
                    outClass = 'pt-page-rotatePushBottom';
                    inClass = 'pt-page-rotatePullTop pt-page-delay180';
                    break;
                case 46:
                    outClass = 'pt-page-rotateFoldLeft';
                    inClass = 'pt-page-moveFromRightFade';
                    break;
                case 47:
                    outClass = 'pt-page-rotateFoldRight';
                    inClass = 'pt-page-moveFromLeftFade';
                    break;
                case 48:
                    outClass = 'pt-page-rotateFoldTop';
                    inClass = 'pt-page-moveFromBottomFade';
                    break;
                case 49:
                    outClass = 'pt-page-rotateFoldBottom';
                    inClass = 'pt-page-moveFromTopFade';
                    break;
                case 50:
                    outClass = 'pt-page-moveToRightFade';
                    inClass = 'pt-page-rotateUnfoldLeft';
                    break;
                case 51:
                    outClass = 'pt-page-moveToLeftFade';
                    inClass = 'pt-page-rotateUnfoldRight';
                    break;
                case 52:
                    outClass = 'pt-page-moveToBottomFade';
                    inClass = 'pt-page-rotateUnfoldTop';
                    break;
                case 53:
                    outClass = 'pt-page-moveToTopFade';
                    inClass = 'pt-page-rotateUnfoldBottom';
                    break;
                case 54:
                    outClass = 'pt-page-rotateRoomLeftOut pt-page-ontop';
                    inClass = 'pt-page-rotateRoomLeftIn';
                    break;
                case 55:
                    outClass = 'pt-page-rotateRoomRightOut pt-page-ontop';
                    inClass = 'pt-page-rotateRoomRightIn';
                    break;
                case 56:
                    outClass = 'pt-page-rotateRoomTopOut pt-page-ontop';
                    inClass = 'pt-page-rotateRoomTopIn';
                    break;
                case 57:
                    outClass = 'pt-page-rotateRoomBottomOut pt-page-ontop';
                    inClass = 'pt-page-rotateRoomBottomIn';
                    break;
                case 58:
                    outClass = 'pt-page-rotateCubeLeftOut pt-page-ontop';
                    inClass = 'pt-page-rotateCubeLeftIn';
                    break;
                case 59:
                    outClass = 'pt-page-rotateCubeRightOut pt-page-ontop';
                    inClass = 'pt-page-rotateCubeRightIn';
                    break;
                case 60:
                    outClass = 'pt-page-rotateCubeTopOut pt-page-ontop';
                    inClass = 'pt-page-rotateCubeTopIn';
                    break;
                case 61:
                    outClass = 'pt-page-rotateCubeBottomOut pt-page-ontop';
                    inClass = 'pt-page-rotateCubeBottomIn';
                    break;
                case 62:
                    outClass = 'pt-page-rotateCarouselLeftOut pt-page-ontop';
                    inClass = 'pt-page-rotateCarouselLeftIn';
                    break;
                case 63:
                    outClass = 'pt-page-rotateCarouselRightOut pt-page-ontop';
                    inClass = 'pt-page-rotateCarouselRightIn';
                    break;
                case 64:
                    outClass = 'pt-page-rotateCarouselTopOut pt-page-ontop';
                    inClass = 'pt-page-rotateCarouselTopIn';
                    break;
                case 65:
                    outClass = 'pt-page-rotateCarouselBottomOut pt-page-ontop';
                    inClass = 'pt-page-rotateCarouselBottomIn';
                    break;
                case 66:
                    outClass = 'pt-page-rotateSidesOut';
                    inClass = 'pt-page-rotateSidesIn pt-page-delay200';
                    break;
                case 67:
                    outClass = 'pt-page-rotateSlideOut';
                    inClass = 'pt-page-rotateSlideIn';
                    break;

            }

            if($current){
                $($current)
                    //.attr( 'farm-select', '' )
                    .addClass(outClass)
                    .on(animEndEventName, function() {
                        //console.log('dentro current ------> ',outClass)
                        $($current)
                            .off(animEndEventName)
                            .removeClass(outClass)
                            .removeAttr('farm-select');
                            //.removeAttribute( 'farm-select' );
                    });

            }

            $($next)
                .attr('farm-select', true)
                //.setAttribute('farm-select', true)
                .addClass(inClass)
                .on(animEndEventName, function() {
                    //console.log('dentro next ------> ',inClass)
                    $($next)
                        .off(animEndEventName)
                        .removeClass(inClass);
                    isAnimating = false;
                });

        //}
    };

    function animation_end(id_page, callback){
        var opts = that.options;

        var animEndEN = animEndEventName();

        if(opts.animation){
            //$("#"+id_page).on("animationend webkitAnimationEnd oAnimationEnd MSAnimationEnd",
            $("#"+id_page).on(animEndEN,
            function(e){
                callback();

                $(this).off(e);
            });
        }
        else{
            callback();
        }
    };



    /////////////////////////////////////////////////////////////////////////////////

    //http://stackoverflow.com/questions/11197247/javascript-equivalent-of-jquerys-extend-method
    function extend(){
        for(var i=1; i<arguments.length; i++)
            for(var key in arguments[i])
                if(arguments[i].hasOwnProperty(key))
                    arguments[0][key] = arguments[i][key];

        return arguments[0];
    }

    return farm_section_bis;
}));
